src/main/java/ch/njol/skript/sections/SecFor.java

```diff
@@ -0,0 +1,154 @@
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.skript.lang.TriggerItem;
+import ch.njol.skript.lang.Variable;
+import ch.njol.skript.lang.util.ContainerExpression;
+import ch.njol.skript.registrations.Feature;
+import ch.njol.skript.util.Container;
+import ch.njol.skript.util.Container.ContainerType;
+import ch.njol.skript.util.LiteralUtils;
+import ch.njol.util.Kleenean;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Map;
+
+@Name("For Each Loop (Experimental)")
+@Description("""
+	A specialised loop section run for each element in a list.
+	Unlike the basic loop, this is designed for extracting the key & value from pairs.
+	The loop element's key/index and value can be stored in a variable for convenience.
+	
+	When looping a simple (non-indexed) set of values, e.g. all players, the index will be the loop counter number."""
+)
+@Examples({
+	"for each {_player} in players:",
+	"\tsend \"Hello %{_player}%!\" to {_player}",
+	"",
+	"loop {_item} in {list of items::*}:",
+	"\tbroadcast {_item}'s name",
+	"",
+	"for each key {_index} in {list of items::*}:",
+	"\tbroadcast {_index}",
+	"",
+	"loop key {_index} and value {_value} in {list of items::*}:",
+	"\tbroadcast \"%{_index}% = %{_value}%\"",
+	"",
+	"for each {_index} = {_value} in {my list::*}:",
+	"\tbroadcast \"%{_index}% = %{_value}%\"",
+})
+@Since("INSERT VERSION")
+public class SecFor extends SecLoop {
+
+	static {
+		Skript.registerSection(SecFor.class,
+			"(for [each]|loop) [value] %~object% in %objects%",
+			"(for [each]|loop) (key|index) %~object% in %objects%",
+			"(for [each]|loop) [key|index] %~object% (=|and) [value] %~object% in %objects%"
+		);
+	}
+
+	private @Nullable Expression<?> keyStore, valueStore;
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public boolean init(Expression<?>[] exprs,
+						int matchedPattern,
+						Kleenean isDelayed,
+						ParseResult parseResult,
+						SectionNode sectionNode,
+						List<TriggerItem> triggerItems) {
+		if (!this.getParser().hasExperiment(Feature.FOR_EACH_LOOPS))
+			return false;
+		//<editor-fold desc="Set the key/value expressions based on the pattern" defaultstate="collapsed">
+		switch (matchedPattern) {
+			case 0:
+				this.valueStore = exprs[0];
+				this.expression = LiteralUtils.defendExpression(exprs[1]);
+				break;
+			case 1:
+				this.keyStore = exprs[0];
+				this.expression = LiteralUtils.defendExpression(exprs[1]);
+				break;
+			default:
+				this.keyStore = exprs[0];
+				this.valueStore = exprs[1];
+				this.expression = LiteralUtils.defendExpression(exprs[2]);
+		}
+		//</editor-fold>
+		//<editor-fold desc="Check our input expressions are safe/correct" defaultstate="collapsed">
+		if (!(keyStore instanceof Variable || keyStore == null)) {
```

Comment:
I think this would be less confusing to instead write as
```suggestion
		if (keyStore != null && !(keyStore instanceof Variable)) {
```
Same for below